osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -325,10 +333,103 @@ private void checkFirmwareHistory(final String deviceId,
+                        e.getValue()))
+                .collect(Collectors.toList());
+
+        // Transform this list so it contains only the latest entry for each
+        // moduleType, using
+        // firmwareVersionsInHistory[x].DeviceFimwareFile.firmwareFile.firmwareModules.firmwareModule.description.
+        // and firmwareVersionsInHistory[x].installationDate
+        Map<String, FirmwareVersionWithInstallationDate> currentlyInstalledFirmwareVersionsPerType = new HashMap<String, FirmwareVersionWithInstallationDate>();
+
+        for (DeviceFirmwareFile firmwareFile : deviceFirmwareFiles) {
+
+            Map<FirmwareModule, String> fwms = firmwareFile.getFirmwareFile().getModuleVersions(); // .keySet();
+            Date installationDate = firmwareFile.getInstallationDate();
+
+            for (Map.Entry<FirmwareModule, String> entry : fwms.entrySet()) {
+                String version = entry.getValue();
+                FirmwareModule fwm = entry.getKey();
+                if (currentlyInstalledFirmwareVersionsPerType.containsKey(fwm.getDescription())) {
+                    // check if this installation of this same kind of module is
+                    // of a later date
+                    if (currentlyInstalledFirmwareVersionsPerType.get(fwm.getDescription()).installationDate
+                            .before(installationDate)) {
+                        currentlyInstalledFirmwareVersionsPerType.replace(fwm.getDescription(),
+                                new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                        FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                    }
+                } else {
+                    // no other module of this type found yet so just add it
+                    currentlyInstalledFirmwareVersionsPerType.put(fwm.getDescription(),
+                            new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                    FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                }
+            }
+        }
+
+        final List<FirmwareVersion> latestfirmwareVersionsOfEachModuleTypeInHistory = currentlyInstalledFirmwareVersionsPerType
+                .values()
+                .stream()
+                .map(e -> new FirmwareVersion(e.firmwareVersion.getFirmwareModuleType(),
+                        e.firmwareVersion.getVersion()))
+                .collect(Collectors.toList());
+
+        // remove the latest history (module)versions from the firmwareVersions
+        // parameter
+        firmwareVersionsToCheck.removeAll(latestfirmwareVersionsOfEachModuleTypeInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    private class FirmwareVersionWithInstallationDate {
+        public Date installationDate;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make installationDate a static final constant or non-public and provide accessors if needed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)